src/java.base/share/classes/java/util/Properties.java

```diff
@@ -903,12 +919,22 @@ private void store0(BufferedWriter bw, String comments, boolean escUnicode)
         if (comments != null) {
             writeComments(bw, comments);
         }
-        bw.write("#" + new Date().toString());
-        bw.newLine();
+        writeDateComment(bw);
+
         synchronized (this) {
-            for (Map.Entry<Object, Object> e : entrySet()) {
-                String key = (String)e.getKey();
-                String val = (String)e.getValue();
+            @SuppressWarnings("unchecked")
+            Collection<Map.Entry<String, String>> entries = (Set<Map.Entry<String, String>>) (Set) entrySet();
+            // entrySet() can be overridden by subclasses. Here we check to see if the returned instance is the one
+            // returned by the Properties.entrySet() implementation.If yes, then we sort those entries
+            // in the natural order of their key. Else we just use the iteration order of the returned instance.
+            if (entries instanceof Collections.SynchronizedSet<?> ss
+                    && ss.c instanceof EntrySet) {
+                entries = new ArrayList<>(entries);
```

Comment:
From the description referring to a 'different instance', I expected to see == or != in this test.
Since Properties.entrySet() always returns a new synchronized set of a new EntrySet,
specifying that the underlying map is the same instance is more difficult.
Perhaps either the SynchronizedSet or the EntrySet instance could be cached and compared.

(typo: missing space after period... ".If yes")

Reply:
A custom subclass couldn't create an instance of EntrySet directly. Therefore if it wants to reorder the entries, it would have to return a new TreeSet or LinkedHashSet - or some other custom set implementation. If the result of calling entrySet() is an EntrySet wrapped in an UnmodifiableSet it's therefore legitimate to assume that the set hasn't been reordered, and that we can reorder it. Or am I missing something?

Reply:
I've updated the PR to clarify what this part of the code is doing. I took Daniel's input to reword the javadoc as well as tried to improve the code comment where this check is done. Essentially, we check for the "type" of the returned intstance and if we see that it's a `SynchronizedSet` whose underlying collection is a private class `EntrySet` then we know that it's the one created by Properties.entrySet(). Like Daniel says, subclasses won't be able to create or use this internal private class and if they did override the entrySet() to change the order of the entries they would have to do it by returning us a different "type" from that method.